From 667eeb78a1401b5e51bf76cdcabedeb4dc0b6ee2 Mon Sep 17 00:00:00 2001
From: Nell <xianbin.zhu@spacemit.com>
Date: Sat, 2 Nov 2024 16:26:52 +0800
Subject: [PATCH 1340/1448] k1:rproc: don't need to stop the threads if they
 ware freezed before

this scenario occurs in power-down process of the hibernation, we may
not be able to stop a thread indefinitely when it has already been
frozen.

Change-Id: Ie83e1e2cb2a5beaaf28639724a60572cdfaa405d
---
 drivers/remoteproc/k1x-rproc.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/drivers/remoteproc/k1x-rproc.c b/drivers/remoteproc/k1x-rproc.c
index c878912c3f56..6b03bce72f84 100644
--- a/drivers/remoteproc/k1x-rproc.c
+++ b/drivers/remoteproc/k1x-rproc.c
@@ -664,7 +664,10 @@ static void spacemit_rproc_shutdown(struct platform_device *pdev)
 
 	for (i = 0; i < MAX_MBOX; ++i) {
 		/* release the resource of rt thread */
-		kthread_stop(priv->mb[i].mb_thread);
+		if (priv->mb[i].kthread_running) {
+			if (!frozen((priv->mb[i].mb_thread)))
+				kthread_stop(priv->mb[i].mb_thread);
+		}
 		/* mbox_free_channel(priv->mb[i].chan); */
 	}
 }
-- 
2.47.0

